src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/PackagePropertyPage/PackageIconValueProvider.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties.Package
+{
+    [ExportInterceptingPropertyValueProvider(PackageIconPropertyName, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class PackageIconValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private const string PackageIconPropertyName = "PackageIcon";
+        private const string PackPropertyName = "Pack";
+        private const string PackagePathPropertyName = "PackagePath";
+        private const string NoneItem = "None";
+        private const string TrueValue = "True";
+
+        private readonly IProjectItemProvider _sourceItemsProvider;
+        private readonly UnconfiguredProject _unconfiguredProject;
+
+        [ImportingConstructor]
+        public PackageIconValueProvider(
+            [Import(ExportContractNames.ProjectItemProviders.SourceFiles)] IProjectItemProvider sourceItemsProvider,
+            UnconfiguredProject unconfiguredProject)
+        {
+            _sourceItemsProvider = sourceItemsProvider;
+            _unconfiguredProject = unconfiguredProject;
+        }
+
+        // https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#packageicon
+        private static string CreatePackageIcon(string filePath, string? packagePath)
+        {
+            string filename = Path.GetFileName(filePath);
+            // Make a slash-only value into empty string, so it won't get prepended onto the path.
+            packagePath = ("\\".Equals(packagePath, StringComparison.Ordinal) || "/".Equals(packagePath, StringComparison.Ordinal))
+                ? string.Empty : (packagePath ?? string.Empty);
+            // The assumption is that PackagePath does not contain a path to a file; only a directory path.
+            return Path.Combine(packagePath, filename);
+        }
+
+        private async Task<IProjectItem?> GetExistingNoneItemAsync(string existingPackageIcon)
+        {
+            foreach (IProjectItem noneItem in await _sourceItemsProvider.GetItemsAsync(NoneItem))
+            {
+                string pack = await noneItem.Metadata.GetEvaluatedPropertyValueAsync(PackPropertyName);
+                string itemIconFilename = Path.GetFileName(noneItem.EvaluatedInclude);
+                string existingIconFilename = Path.GetFileName(existingPackageIcon);
```

Comment:
These calls will allocate strings for every none item in the project. If you check the `pack` value first, you can avoid those allocations for all items that do not have `Pack="true"`.

Reply:
I had that at first. After a few iterations, I realized I didn't need to have 2 consecutive if-statements. This is done strictly for human readability, because you can drop these method calls in-place in the predicate for the if-statement, and they'll only get executed as-needed because of the short-circuiting of the && operation. But if that's what is desired, I'll do that. Fixed.

Reply:
In general we should avoid allocating memory for no reason.